qiskit/pulse/basepasses.py

```diff
@@ -0,0 +1,176 @@
+from inspect import signature
+from typing import Any
+
+from qiskit import pulse
+from qiskit.pulse.states import Analysis, State
+
+
+class MetaPass(type):
+    """Metaclass for pulse compilation passes.
+
+    Enforces the creation of some fields in the pass while allowing passes to
+    override ``__init__``.
+    """
+
+    def __call__(cls, *args, **kwargs):
+        pass_instance = type.__call__(cls, *args, **kwargs)
+        pass_instance._hash = hash(MetaPass._freeze_init_parameters(cls, args, kwargs))
+        return pass_instance
+
+    @staticmethod
+    def _freeze_init_parameters(class_, args, kwargs):
+        self_guard = object()
+        init_signature = signature(class_.__init__)
+        bound_signature = init_signature.bind(self_guard, *args, **kwargs)
+        arguments = [('class_.__name__', class_.__name__)]
+        for name, value in bound_signature.arguments.items():
+            if value == self_guard:
+                continue
+            if isinstance(value, Hashable):
+                arguments.append((name, type(value), value))
+            else:
+                arguments.append((name, type(value), repr(value)))
+        return frozenset(arguments)
+
+
+class BasePass(metaclass=MetaPass):
+    """Base class for transpiler passes."""
+
+    def __init__(self):
+        self.requires = []  # List of passes that requires
+        self.preserves = []  # List of passes that preserves
+        self.state = State()
+        self._hash = None
+
+    def __hash__(self):
+        return self._hash
+
+    def __eq__(self, other):
+        return hash(self) == hash(other)
+
+    @property
+    def name(self):
+        """Return the name of the pass."""
+        return self.__class__.__name__
+
+    @property
+    def analysis(self) -> Analysis:
+        """Return the current analysis information available to this pass."""
+        return self.state.analysis
+
+    @abstractmethod
+    def run(self, program: pulse.Program) -> pulse.Program:
+        """Run a pass on the pulse program. This is implemented by the pass developer.
+
+        Args:
+            program: The program on which the pass is run.
+        Raises:
+            NotImplementedError: when this is left unimplemented for a pass.
+        """
+        raise NotImplementedError
+
+    @property
+    def is_transformation_pass(self):
+        """Check if the pass is a transformation pass."""
+        return isinstance(self, TransformationPass)
+
+    @property
+    def is_analysis_pass(self):
+        """Check if the pass is an analysis pass."""
+        return isinstance(self, AnalysisPass)
+
+    @property
+    def is_lowering_pass(self):
+        """Check if the pass is a lowering pass."""
+        return isinstance(self, LoweringPass)
+
+    @property
+    def is_validation_pass(self):
+        """Check if the pass is a validation pass."""
+        return isinstance(self, ValidationPass)
+
+
+class AnalysisPass(BasePass):  # pylint: disable=abstract-method
```

Comment:
Why isn't this in `analysis.py`?

Reply:
This is the format of `basepasses` in the transpiler, which contains the base abstract class definitions. `analysis` is meant for implementations, or as you note above we can put them in `passes.analysis`, `passes.transforms`, etc.